File: roaming/csharp/Json.cs
// JSON for roam: a reader with cleanroom-transformer's grammar (src/json.c), working on UTF-8 bytes,
// and a writer in Python's json.dumps(ensure_ascii=False) style. Mirrors ../node/json.mjs.
using System;
using System.Collections.Generic;
using System.Numerics;
using System.Text;

namespace SemIf.Roam;

public enum JType { Null, True, False, Int, Float, String, Array, Object }

public sealed class JValue
{
    public JType Type;
    public string Digits = "";          // Int, as written ("-0" normalized to "0")
    public byte[] Bytes = Array.Empty<byte>(); // String (UTF-8)
    public string Str = "";             // String
    public List<JValue> Items = new();  // Array items, Object values
    public List<string> Keys = new();   // Object keys, in order, duplicates kept

    /// <summary>The value of the last pair named <paramref name="key"/> (as json_get sees it).</summary>
    public JValue? Get(string key)
    {
        if (Type != JType.Object) return null;
        for (int k = Keys.Count - 1; k >= 0; k--) if (Keys[k] == key) return Items[k];
        return null;
    }

    /// <summary>An object with exactly these keys, each once.</summary>
    public bool HasExactKeys(params string[] keys)
    {
        if (Type != JType.Object || Keys.Count != keys.Length) return false;
        foreach (var k in keys)
        {
            int n = 0;
            foreach (var x in Keys) if (x == k) n++;
            if (n != 1) return false;
        }
        return true;
    }

    /// <summary>A non-negative integer up to 2^64 - 1 (parse_u64), else null.</summary>
    public BigInteger? U64()
    {
        if (Type != JType.Int || Digits.StartsWith('-')) return null;
        var x = BigInteger.Parse(Digits);
        return x <= ulong.MaxValue ? x : null;
    }
}

public sealed class JsonException : Exception
{
    public JsonException(string m) : base(m) { }
}

public static class Json
{
    public static readonly UTF8Encoding Strict = new(false, true);
    const int MaxDepth = 512;

    public static bool Utf8Valid(ReadOnlySpan<byte> s)
    {
        try { Strict.GetCharCount(s); return true; }
        catch (DecoderFallbackException) { return false; }
    }

    public static JValue Parse(ReadOnlySpan<byte> input) => new Parser(input.ToArray()).Run();

    sealed class Parser
    {
        readonly byte[] s;
        int i, depth;
        public Parser(byte[] s) { this.s = s; }
        int N => s.Length;

        Exception Fail(string what) => new JsonException($"{what} at byte {i}");

        void Ws() { while (i < N && (s[i] == 0x20 || s[i] == 0x09 || s[i] == 0x0a || s[i] == 0x0d)) i++; }

        bool Lit(string w)
        {
            if (N - i < w.Length) return false;
            for (int k = 0; k < w.Length; k++) if (s[i + k] != w[k]) return false;
            i += w.Length;
            return true;
        }

        uint Hex4()
        {
            if (N - i < 4) throw Fail("short \\u escape");
            uint v = 0;
            for (int k = 0; k < 4; k++)
            {
                byte c = s[i++];
                v <<= 4;
                if (c >= '0' && c <= '9') v |= (uint)(c - '0');
                else if (c >= 'a' && c <= 'f') v |= (uint)(c - 'a' + 10);
                else if (c >= 'A' && c <= 'F') v |= (uint)(c - 'A' + 10);
                else throw Fail("bad \\u escape");
            }
            return v;
        }

        JValue String()
        {
            i++;
            var b = new List<byte>();
            int run = i;
            for (;;)
            {
                if (i >= N) throw Fail("unterminated string");
                byte c = s[i];
                if (c == '"') { for (int k = run; k < i; k++) b.Add(s[k]); i++; break; }
                if (c < 0x20) throw Fail("control character in string");
                if (c != '\\') { i++; continue; }
                for (int k = run; k < i; k++) b.Add(s[k]);
                i++;
                if (i >= N) throw Fail("bad escape");
                byte e = s[i++];
                switch (e)
                {
                    case (byte)'"': b.Add((byte)'"'); break;
                    case (byte)'\\': b.Add((byte)'\\'); break;
                    case (byte)'/': b.Add((byte)'/'); break;
                    case (byte)'b': b.Add(8); break;
                    case (byte)'f': b.Add(12); break;
                    case (byte)'n': b.Add(10); break;
                    case (byte)'r': b.Add(13); break;
                    case (byte)'t': b.Add(9); break;
                    case (byte)'u':
                    {
                        uint cp = Hex4();
                        if (cp >= 0xD800 && cp <= 0xDBFF)
                        {
                            if (N - i < 6 || s[i] != '\\' || s[i + 1] != 'u') throw Fail("lone surrogate");
                            i += 2;
                            uint lo = Hex4();
                            if (lo < 0xDC00 || lo > 0xDFFF) throw Fail("lone surrogate");
                            cp = 0x10000 + ((cp - 0xD800) << 10) + (lo - 0xDC00);
                        }
                        else if (cp >= 0xDC00 && cp <= 0xDFFF) throw Fail("lone surrogate");
                        b.AddRange(Encoding.UTF8.GetBytes(char.ConvertFromUtf32((int)cp)));
                        break;
                    }
                    default: throw Fail("bad escape");
                }
                run = i;
            }
            var bytes = b.ToArray();
            return new JValue { Type = JType.String, Bytes = bytes, Str = Strict.GetString(bytes) };
        }

        JValue Number()
        {
            int start = i;
            bool isFloat = false;
            if (s[i] == '-') i++;
            if (i < N && s[i] == '0') i++;
            else if (i < N && s[i] >= '1' && s[i] <= '9') { while (i < N && s[i] >= '0' && s[i] <= '9') i++; }
            else
            {
                if (Lit("Infinity")) return new JValue { Type = JType.Float };
                throw Fail("bad number");
            }
            if (i < N && s[i] == '.')
            {
                int d = ++i;
                while (i < N && s[i] >= '0' && s[i] <= '9') i++;
                if (i == d) throw Fail("bad fraction");
                isFloat = true;
            }
            if (i < N && (s[i] == 'e' || s[i] == 'E'))
            {
                i++;
                if (i < N && (s[i] == '+' || s[i] == '-')) i++;
                int d = i;
                while (i < N && s[i] >= '0' && s[i] <= '9') i++;
                if (i == d) throw Fail("bad exponent");
                isFloat = true;
            }
            if (isFloat) return new JValue { Type = JType.Float };
            var digits = Encoding.ASCII.GetString(s, start, i - start);
            if (digits == "-0") digits = "0";
            return new JValue { Type = JType.Int, Digits = digits };
        }

        JValue Value()
        {
            if (++depth > MaxDepth) throw Fail("nesting too deep");
            Ws();
            if (i >= N) throw Fail("unexpected end");
            byte c = s[i];
            JValue v;
            if (c == '{')
            {
                i++;
                v = new JValue { Type = JType.Object };
                Ws();
                if (i < N && s[i] == '}') i++;
                else
                    for (;;)
                    {
                        Ws();
                        if (i >= N || s[i] != '"') throw Fail("expected string key");
                        var k = String();
                        Ws();
                        if (i >= N || s[i] != ':') throw Fail("expected :");
                        i++;
                        v.Keys.Add(k.Str);
                        v.Items.Add(Value());
                        Ws();
                        if (i < N && s[i] == ',') { i++; continue; }
                        if (i < N && s[i] == '}') { i++; break; }
                        throw Fail("expected , or }");
                    }
            }
            else if (c == '[')
            {
                i++;
                v = new JValue { Type = JType.Array };
                Ws();
                if (i < N && s[i] == ']') i++;
                else
                    for (;;)
                    {
                        v.Items.Add(Value());
                        Ws();
                        if (i < N && s[i] == ',') { i++; continue; }
                        if (i < N && s[i] == ']') { i++; break; }
                        throw Fail("expected , or ]");
                    }
            }
            else if (c == '"') v = String();
            else if (c == '-' || (c >= '0' && c <= '9')) v = Number();
            else if (Lit("true")) v = new JValue { Type = JType.True };
            else if (Lit("false")) v = new JValue { Type = JType.False };
            else if (Lit("null")) v = new JValue { Type = JType.Null };
            else if (Lit("NaN") || Lit("Infinity")) v = new JValue { Type = JType.Float };
            else throw Fail("unexpected character");
            depth--;
            return v;
        }

        public JValue Run()
        {
            if (!Utf8Valid(s)) throw new JsonException("input is not valid UTF-8");
            var v = Value();
            Ws();
            if (i != N) throw Fail("trailing data");
            return v;
        }
    }

    // ---------------------------------------------------------------- writer
    public static string DumpStr(string s)
    {
        var sb = new StringBuilder("\"");
        foreach (var r in s.EnumerateRunes())
        {
            int c = r.Value;
            switch (c)
            {
                case '"': sb.Append("\\\""); break;
                case '\\': sb.Append("\\\\"); break;
                case '\n': sb.Append("\\n"); break;
                case '\r': sb.Append("\\r"); break;
                case '\t': sb.Append("\\t"); break;
                case '\b': sb.Append("\\b"); break;
                case '\f': sb.Append("\\f"); break;
                default:
                    if (c < 0x20) sb.Append("\\u00").Append(c.ToString("x2"));
                    else sb.Append(r.ToString());
                    break;
            }
        }
        return sb.Append('"').ToString();
    }

    /// <summary>Writes string, integer (BigInteger/long/int/ulong), bool, List of values and
    /// ordered objects (List of key/value pairs).</summary>
    public static string Dump(object v) => v switch
    {
        string s => DumpStr(s),
        bool b => b ? "true" : "false",
        BigInteger n => n.ToString(),
        int n => n.ToString(),
        long n => n.ToString(),
        ulong n => n.ToString(),
        Obj o => "{" + string.Join(", ", o.Pairs.ConvertAll(p => DumpStr(p.Key) + ": " + Dump(p.Value))) + "}",
        List<object> l => "[" + string.Join(", ", l.ConvertAll(Dump)) + "]",
        _ => throw new ArgumentException($"cannot dump {v.GetType()}"),
    };
}

/// <summary>A JSON object that keeps its key order.</summary>
public sealed class Obj
{
    public List<KeyValuePair<string, object>> Pairs = new();
    public Obj Add(string k, object v) { Pairs.Add(new(k, v)); return this; }
}

File: roaming/node/json.mjs
// JSON for roam: a reader with cleanroom-transformer's grammar (src/json.c) and a writer in
// Python's json.dumps(ensure_ascii=False) style. Both work on UTF-8 bytes, so Node and the C#
// port accept and produce exactly the same text.

const dec = new TextDecoder('utf-8', { fatal: true, ignoreBOM: true }); // keep a leading U+FEFF as data
const enc = new TextEncoder();
const MAX_DEPTH = 512;

export class JsonError extends Error {}

function utf8Valid(s) {
  try {
    dec.decode(s);
  } catch {
    return false;
  }
  return true; // TextDecoder(fatal) rejects overlongs, surrogates and > U+10FFFF, like utf8_valid
}

/**
 * Parses UTF-8 bytes. Values: {t: 'null'|'true'|'false'|'int'|'float'|'string'|'array'|'object', ...}
 * int: {digits} (as written, "-0" normalized to "0"); string: {s, bytes}; array: {items};
 * object: {keys: string[], vals} (all pairs, in order, duplicates kept).
 */
export function parseJson(bytes) {
  if (!utf8Valid(bytes)) throw new JsonError('input is not valid UTF-8');
  let i = 0, depth = 0;
  const n = bytes.length;
  const fail = (what) => {
    throw new JsonError(`${what} at byte ${i}`);
  };
  const ws = () => {
    while (i < n && (bytes[i] === 0x20 || bytes[i] === 0x09 || bytes[i] === 0x0a || bytes[i] === 0x0d)) i++;
  };
  const lit = (word) => {
    for (let k = 0; k < word.length; k++) if (bytes[i + k] !== word.charCodeAt(k)) return false;
    i += word.length;
    return true;
  };
  const hex4 = () => {
    if (n - i < 4) fail('short \\u escape');
    let v = 0;
    for (let k = 0; k < 4; k++) {
      const c = bytes[i++];
      v <<= 4;
      if (c >= 0x30 && c <= 0x39) v |= c - 0x30;
      else if (c >= 0x61 && c <= 0x66) v |= c - 0x61 + 10;
      else if (c >= 0x41 && c <= 0x46) v |= c - 0x41 + 10;
      else fail('bad \\u escape');
    }
    return v;
  };
  const string = () => {
    i++;
    const out = [];
    let run = i;
    for (;;) {
      if (i >= n) fail('unterminated string');
      const c = bytes[i];
      if (c === 0x22) {
        out.push(bytes.subarray(run, i));
        i++;
        break;
      }
      if (c < 0x20) fail('control character in string');
      if (c !== 0x5c) {
        i++;
        continue;
      }
      out.push(bytes.subarray(run, i));
      i++;
      if (i >= n) fail('bad escape');
      const e = bytes[i++];
      const simple = { 0x22: 0x22, 0x5c: 0x5c, 0x2f: 0x2f, 0x62: 8, 0x66: 12, 0x6e: 10, 0x72: 13, 0x74: 9 }[e];
      if (simple !== undefined) out.push(Uint8Array.of(simple));
      else if (e === 0x75) {
        let cp = hex4();
        if (cp >= 0xd800 && cp <= 0xdbff) {
          if (n - i < 6 || bytes[i] !== 0x5c || bytes[i + 1] !== 0x75) fail('lone surrogate');
          i += 2;
          const lo = hex4();
          if (lo < 0xdc00 || lo > 0xdfff) fail('lone surrogate');
          cp = 0x10000 + ((cp - 0xd800) << 10) + (lo - 0xdc00);
        } else if (cp >= 0xdc00 && cp <= 0xdfff) fail('lone surrogate');
        out.push(enc.encode(String.fromCodePoint(cp)));
      } else fail('bad escape');
      run = i;
    }
    const total = out.reduce((a, p) => a + p.length, 0);
    const b = new Uint8Array(total);
    let at = 0;
    for (const p of out) b.set(p, at), (at += p.length);
    return { t: 'string', bytes: b, s: dec.decode(b) };
  };
  const number = () => {
    const start = i;
    let isFloat = false;
    if (bytes[i] === 0x2d) i++;
    if (i < n && bytes[i] === 0x30) i++;
    else if (i < n && bytes[i] >= 0x31 && bytes[i] <= 0x39) while (i < n && bytes[i] >= 0x30 && bytes[i] <= 0x39) i++;
    else {
      if (lit('Infinity')) return { t: 'float' };
      fail('bad number');
    }
    if (i < n && bytes[i] === 0x2e) {
      const d = ++i;
      while (i < n && bytes[i] >= 0x30 && bytes[i] <= 0x39) i++;
      if (i === d) fail('bad fraction');
      isFloat = true;
    }
    if (i < n && (bytes[i] === 0x65 || bytes[i] === 0x45)) {
      i++;
      if (i < n && (bytes[i] === 0x2b || bytes[i] === 0x2d)) i++;
      const d = i;
      while (i < n && bytes[i] >= 0x30 && bytes[i] <= 0x39) i++;
      if (i === d) fail('bad exponent');
      isFloat = true;
    }
    if (isFloat) return { t: 'float' };
    let digits = String.fromCharCode(...bytes.subarray(start, i));
    if (digits === '-0') digits = '0';
    return { t: 'int', digits };
  };
  const value = () => {
    if (++depth > MAX_DEPTH) fail('nesting too deep');
    ws();
    if (i >= n) fail('unexpected end');
    const c = bytes[i];
    let v;
    if (c === 0x7b) {
      i++;
      const keys = [], vals = [];
      ws();
      if (i < n && bytes[i] === 0x7d) i++;
      else
        for (;;) {
          ws();
          if (i >= n || bytes[i] !== 0x22) fail('expected string key');
          const k = string();
          ws();
          if (i >= n || bytes[i] !== 0x3a) fail('expected :');
          i++;
          keys.push(k.s);
          vals.push(value());
          ws();
          if (i < n && bytes[i] === 0x2c) {
            i++;
            continue;
          }
          if (i < n && bytes[i] === 0x7d) {
            i++;
            break;
          }
          fail('expected , or }');
        }
      v = { t: 'object', keys, vals };
    } else if (c === 0x5b) {
      i++;
      const items = [];
      ws();
      if (i < n && bytes[i] === 0x5d) i++;
      else
        for (;;) {
          items.push(value());
          ws();
          if (i < n && bytes[i] === 0x2c) {
            i++;
            continue;
          }
          if (i < n && bytes[i] === 0x5d) {
            i++;
            break;
          }
          fail('expected , or ]');
        }
      v = { t: 'array', items };
    } else if (c === 0x22) v = string();
    else if (c === 0x2d || (c >= 0x30 && c <= 0x39)) v = number();
    else if (lit('true')) v = { t: 'true' };
    else if (lit('false')) v = { t: 'false' };
    else if (lit('null')) v = { t: 'null' };
    else if (lit('NaN') || lit('Infinity')) v = { t: 'float' };
    else fail('unexpected character');
    depth--;
    return v;
  };
  const v = value();
  ws();
  if (i !== n) fail('trailing data');
  return v;
}

/** The value of `key` in an object (the last one, as json_get sees it), or undefined. */
export function get(obj, key) {
  if (!obj || obj.t !== 'object') return undefined;
  for (let k = obj.keys.length - 1; k >= 0; k--) if (obj.keys[k] === key) return obj.vals[k];
  return undefined;
}

/** An object whose keys are exactly `keys`, once each and in any order. */
export function hasExactKeys(obj, keys) {
  return obj?.t === 'object' && obj.keys.length === keys.length && keys.every((k) => obj.keys.filter((x) => x === k).length === 1);
}

/** Non-negative integer without overflow past 2^64 - 1, as parse_u64. */
export function u64(v) {
  if (!v || v.t !== 'int' || v.digits.startsWith('-')) return undefined;
  const x = BigInt(v.digits);
  return x <= 0xffffffffffffffffn ? x : undefined;
}

// ---------------------------------------------------------------- writer
export function dumpStr(s) {
  let out = '"';
  for (const ch of s) {
    const c = ch.codePointAt(0);
    if (ch === '"') out += '\\"';
    else if (ch === '\\') out += '\\\\';
    else if (ch === '\n') out += '\\n';
    else if (ch === '\r') out += '\\r';
    else if (ch === '\t') out += '\\t';
    else if (ch === '\b') out += '\\b';
    else if (ch === '\f') out += '\\f';
    else if (c < 0x20) out += `\\u00${c.toString(16).padStart(2, '0')}`;
    else out += ch;
  }
  return `${out}"`;
}

/** Writes strings, bigints/integers, booleans, arrays and plain objects (key order kept). */
export function dump(v) {
  if (typeof v === 'string') return dumpStr(v);
  if (typeof v === 'bigint' || typeof v === 'number') return String(v);
  if (typeof v === 'boolean') return v ? 'true' : 'false';
  if (Array.isArray(v)) return `[${v.map(dump).join(', ')}]`;
  return `{${Object.entries(v).map(([k, x]) => `${dumpStr(k)}: ${dump(x)}`).join(', ')}}`;
}
